### PR TITLE
Make path to append_metadata.cmake relative to top-level CMakeLists.txt

### PR DESCRIPTION
**Pull request by Flogex**
Follow-up to https://github.com/duckdb/duckdb/commit/0ecc6820ee6c9b3f2fac2713eb253393fe97bdf2:
`CMAKE_SOURCE_DIR` is used to specify the location of the _scripts_ directory. However, this variable does not point to the duckdb top-level directory if DuckDB is built from within another CMake project, e.g. for extensions. I replaced the variable with `CMAKE_CURRENT_FUNCTION_LIST_DIR`which gives the directory of the CMakeLists.txt where the `build_loadable_extension_directory` function is defined—in other words, the top-level directory.

Similarly, when looking for the `duckdb_platform_out` file, `CMAKE_BINARY_DIR` is the wrong path when DuckDB is not the top-level CMake project. I replaced it with `DuckDB_BINARY_DIR` where "DuckDB" is the [`PROJECT-NAME`](https://cmake.org/cmake/help/latest/variable/PROJECT-NAME_BINARY_DIR.html).

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -820,7 +820,7 @@ function(build_loadable_extension_directory NAME OUTPUT_DIRECTORY EXTENSION_VERS
     TARGET ${TARGET_NAME}
     POST_BUILD
     COMMAND
-    ${CMAKE_COMMAND} -DEXTENSION=$<TARGET_FILE:${TARGET_NAME}> -DPLATFORM_FILE=${CMAKE_BINARY_DIR}/duckdb_platform_out -DDUCKDB_VERSION="${DUCKDB_NORMALIZED_VERSION}" -DEXTENSION_VERSION="${EXTENSION_VERSION}" -DNULL_FILE=${CMAKE_SOURCE_DIR}/scripts/null.txt -P ${CMAKE_SOURCE_DIR}/scripts/append_metadata.cmake
+    ${CMAKE_COMMAND} -DEXTENSION=$<TARGET_FILE:${TARGET_NAME}> -DPLATFORM_FILE=${DuckDB_BINARY_DIR}/duckdb_platform_out -DDUCKDB_VERSION="${DUCKDB_NORMALIZED_VERSION}" -DEXTENSION_VERSION="${EXTENSION_VERSION}" -DNULL_FILE=${CMAKE_CURRENT_FUNCTION_LIST_DIR}/scripts/null.txt -P ${CMAKE_CURRENT_FUNCTION_LIST_DIR}/scripts/append_metadata.cmake
```
Comment (carlopi) [inline CMakeLists.txt:823]:
```suggestion
    ${CMAKE_COMMAND} -DEXTENSION=$<TARGET_FILE:${TARGET_NAME}> -DPLATFORM_FILE=${PROJECT_BINARY_DIR}/duckdb_platform_out -DDUCKDB_VERSION="${DUCKDB_NORMALIZED_VERSION}" -DEXTENSION_VERSION="${EXTENSION_VERSION}" -DNULL_FILE=${CMAKE_CURRENT_FUNCTION_LIST_DIR}/scripts/null.txt -P ${CMAKE_CURRENT_FUNCTION_LIST_DIR}/scripts/append_metadata.cmake
```
Would PROJECT_BINARY_DIR work? That is currently used for determining the location of duckdb_platform_binary in:
```cmake
set_target_properties(duckdb_platform_binary PROPERTIES RUNTIME_OUTPUT_DIRECTORY
                                         ${PROJECT_BINARY_DIR})
```
and implicitly for the location of `duckdb_platform_out`.

I think it should work the same, and avoid the naming that looks more confusing to me.

**Review by carlopi — COMMENTED** [review]
Thanks a lot for filing this correcting the limitations introduced.

Could you check whether the proposed change works the same? If so, I would prefer the more generic PROJECT_BINARY_DIR given it's used already elsewhere in the CMake logic.

One additional request, independent from the PR: would it be possible to submit a demo CMake that showcases the setup where this was a problem? Then that can be used as part of CI to check modifying the CMake logic is not impacting your setup.
I can help move that into a CI check, but even just a CMake with a add_subproject or so would be cool to have, possibly in an examples folder.

Thanks!

**Comment by Flogex** [comment]
> One additional request, independent from the PR: would it be possible to submit a demo CMake that showcases the setup where this was a problem? Then that can be used as part of CI to check modifying the CMake logic is not impacting your setup.
I can help move that into a CI check, but even just a CMake with a add_subproject or so would be cool to have, possibly in an examples folder.

Hey @carlopi. I've built a repro using the extension template. But instead of using the `build_loadable_extension`/`build_static_extension` functions, duckdb is added as subdirectory and the duckdb targets linked against the extension.
https://github.com/Flogex/duckdb-extension-template/commit/7aa87de2b2cf7c7160a04ffdb2f16c623e9fae52

Now when running `make release`, the `CMAKE_BINARY_DIR` is _extension-template/build/Release/_ and the `CMAKE_SOURCE_DIR` is _extension-template/_ (as set in the _release_ make target).
CMake will fail with:
> CMake Error: Not a file: extension-template/scripts/append_metadata.cmake

Let me know if this is helpful.

File: CMakeLists.txt (hunk 1 above)
Comment (Flogex) [reply CMakeLists.txt:823]: `PROJECT_BINARY_DIR` will not work when the `build_loadable_extension` function is called from another CMake project. For example, for the Parquet extension:
https://github.com/duckdb/duckdb/blob/e0c4d9c4dcb08573dba54df18a460ad8606ac9a0/extension/parquet/CMakeLists.txt#L68

The `PROJECT_BINARY_DIR` inside the function would be _build/release/extension/parquet_, not _build/release_.

Comment (carlopi) [reply CMakeLists.txt:823]: Thanks for double checking.

**Review by carlopi — APPROVED** [review]

**Comment by carlopi** [comment]
This is good to go for me.
Thanks @Flogex for sharing also the demo-setup, I will add something inspired by that in the CI, so to flag those problems earlier.
I will also a setup similar to original extension-template. That was working in this case (since I tested it), but it's not a given that any change has to be compatible.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**